outbox.js

```diff
@@ -0,0 +1,219 @@
+      }
+      // else {
+      //   console.log(`Duplicate item with key: ${itemKey} skipped`);
+      // }
+    }
+  }
+
+  async *fetchOutboxItems(outboxUrl) {
+    if (!outboxUrl) {
+      console.error("No outbox URL provided");
+      return;
+    }
+
+    try {
+      const response = await fetch(outboxUrl);
+      if (!response.ok) {
+        throw new Error(`HTTP error! Status: ${response.status}`);
+      }
+      const outboxData = await response.json();
+
+      this.totalPages = Math.ceil(
+        outboxData.orderedItems.length / this.numPosts
+      );
+      // Prevent page number from going beyond total pages
+      this.page = Math.min(this.page, this.totalPages);
+
+      // Simulate pagination by slicing the items array
+      const startIndex = (this.page - 1) * this.numPosts;
+      const endIndex = startIndex + this.numPosts;
+      const paginatedItems = outboxData.orderedItems.slice(
+        startIndex,
+        endIndex
+      );
+
+      for (const item of paginatedItems) {
+        yield item;
+      }
+    } catch (error) {
+      console.error("Error fetching outbox:", error);
+    }
+  }
+
+  renderItem(item) {
+    const activityElement = document.createElement("distributed-activity");
+    activityElement.type = item.type;
+    activityElement.data = item;
+    this.appendChild(activityElement);
+  }
+
+  paginationControls() {
+    // Attach event listeners for pagination
+    const prevButton = document.getElementById("prevPage");
+    const nextButton = document.getElementById("nextPage");
+    if (prevButton) {
+      prevButton.addEventListener("click", () => this.prevPage());
+    }
+    if (nextButton) {
+      nextButton.addEventListener("click", () => this.nextPage());
+    }
+  }
+
+  nextPage() {
+    const currentPage = this.page;
+    if (currentPage < this.totalPages) {
+      this.setAttribute("page", currentPage + 1);
+    }
+  }
+
+  prevPage() {
+    const currentPage = this.page;
+    this.setAttribute("page", Math.max(1, currentPage - 1));
+  }
+
+  attributeChangedCallback(name, oldValue, newValue) {
+    if (name === "url") {
+      this.clearContent();
+      this.loadOutbox(newValue);
+    } else if (name === "num-posts" || name === "page") {
+      // Convert attribute name from kebab-case to camelCase
+      const propName = name.replace(/-([a-z])/g, (g) => g[1].toUpperCase());
+      this[propName] = parseInt(newValue, 10);
+      this.clearContent();
+      this.loadOutbox(this.getAttribute("url"));
+    }
+  }
+
+  clearContent() {
+    // Clear existing content
+    this.innerHTML = "";
+    this.renderedItems.clear();
+  }
+}
+
+// Register the new element with the browser
+customElements.define("distributed-outbox", DistributedOutbox);
+
+class DistributedActivity extends HTMLElement {
+  constructor() {
+    super();
+    this.activityType = "";
+    this.activityData = {};
+  }
+
+  static get observedAttributes() {
+    return ["type", "data"];
+  }
+
+  connectedCallback() {
```

Comment:
We should also support loading from a `url` in case these properties are missing